// sec-edgar-api/build/services/DocumentParser/DocumentParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var XMLParser_1 = require("./XMLParser");
var parsers_1 = require("./parsers");
var DocumentParser = /** @class */ (function () {
    function DocumentParser(args) {
        var _a = args !== null && args !== void 0 ? args : {}, _b = _a.parser, parser = _b === void 0 ? new XMLParser_1.default() : _b, _c = _a.parsersByName, parsersByName = _c === void 0 ? parsers_1.default : _c;
        this.parser = parser;
        this.parsersByName = parsersByName;
    }
    DocumentParser.prototype.parseForm4 = function (params) {
        return this.parsersByName.parseForm4(params, this.parser);
    };
    DocumentParser.prototype.parseForm13g = function (params) {
        return this.parsersByName.parseForm13g(params, this.parser);
    };
    DocumentParser.prototype.parseForm10k = function (params) {
        return this.parsersByName.parseForm10k(params, this.parser);
    };
    DocumentParser.prototype.parseFormDef14a = function (params) {
        return this.parsersByName.parseFormDef14a(params, this.parser);
    };
    return DocumentParser;
}());
exports.default = DocumentParser;
