// sec-edgar-api/build/services/ReportParser/resolvers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var resolve_asset_current_1 = require("./resolve-asset-current");
var resolve_asset_non_current_ppe_gross_1 = require("./resolve-asset-non-current-ppe-gross");
var resolve_cash_flow_capex_1 = require("./resolve-cash-flow-capex");
var resolve_cash_flow_free_1 = require("./resolve-cash-flow-free");
var resolve_cash_flow_operating_1 = require("./resolve-cash-flow-operating");
var resolve_cash_flow_working_capital_non_cash_1 = require("./resolve-cash-flow-working-capital-non-cash");
var resolve_ebit_1 = require("./resolve-ebit");
var resolve_expense_depreciation_1 = require("./resolve-expense-depreciation");
var resolve_expense_operating_1 = require("./resolve-expense-operating");
var resolve_expense_total_1 = require("./resolve-expense-total");
var resolve_fiscal_year_cumulative_properties_1 = require("./resolve-fiscal-year-cumulative-properties");
var resolve_liability_current_1 = require("./resolve-liability-current");
var resolve_q4_fiscal_year_matching_properties_1 = require("./resolve-q4-fiscal-year-matching-properties");
var resolve_revenue_total_1 = require("./resolve-revenue-total");
var resolvers = {
    resolveAssetNonCurrentPPEGross: resolve_asset_non_current_ppe_gross_1.resolveAssetNonCurrentPPEGross,
    resolveExpenseDepreciation: resolve_expense_depreciation_1.resolveExpenseDepreciation,
    resolveExpenseOperating: resolve_expense_operating_1.resolveExpenseOperating,
    resolveCashFlowFree: resolve_cash_flow_free_1.resolveCashFlowFree,
    resolveEbit: resolve_ebit_1.resolveEbit,
    resolveAssetCurrent: resolve_asset_current_1.resolveAssetCurrent,
    resolveCashFlowCapex: resolve_cash_flow_capex_1.resolveCashFlowCapex,
    resolveFiscalYearCumulativeProperties: resolve_fiscal_year_cumulative_properties_1.resolveFiscalYearCumulativeProperties,
    resolveRevenueTotal: resolve_revenue_total_1.resolveRevenueTotal,
    resolveExpenseTotal: resolve_expense_total_1.resolveExpenseTotal,
    resolveCashFlowOperating: resolve_cash_flow_operating_1.resolveCashFlowOperating,
    resolveLiabilityCurrent: resolve_liability_current_1.resolveLiabilityCurrent,
    resolveQ4FiscalYearMatchingProperties: resolve_q4_fiscal_year_matching_properties_1.resolveQ4FiscalYearMatchingProperties,
    resolveCashFlowWorkingCapitalNonCash: resolve_cash_flow_working_capital_non_cash_1.resolveCashFlowWorkingCapitalNonCash,
};
exports.default = resolvers;
